test/select.jl

```diff
@@ -720,20 +720,32 @@ end
         @test_throws ArgumentError select(df, :x => x -> retval)
         @test_throws ArgumentError select(df, :x => x -> retval, copycols=false)
         @test_throws ArgumentError select!(df, :x => x -> retval)
+
+        @test select(df, :x => ByRow(x -> retval)) == DataFrame(x_function = [retval])
+        cdf = copy(df)
+        select!(cdf, :x => ByRow(x -> retval))
+        @test cdf == DataFrame(x_function = [retval])
+
         if retval isa Union{NamedTuple, DataFrameRow}
-            @test_throws ArgumentError select(df, :x => ByRow(x -> retval))
-            @test_throws ArgumentError select!(df, :x => ByRow(x -> retval))
-        else
-            @test select(df, :x => ByRow(x -> retval)) == DataFrame(x_function = [retval])
-            cdf = copy(df)
-            select!(cdf, :x => ByRow(x -> retval))
-            @test cdf == DataFrame(x_function = [retval])
+            @test select(df, :x => ByRow(x -> retval) => AsTable) == DataFrame(;retval...)
+        elseif retval isa DataFrame
+            @test_throws MethodError select(df, :x => ByRow(x -> retval) => AsTable)
+        else # Matrix; surprising but following the API
+            @test select(df, :x => ByRow(x -> retval) => AsTable) ==
+                  DataFrame(["CartesianIndex($i, $j)" => 1.0 for i in 1:2, j in 1:2]...)
```

Comment:
Can you comment on what happens here? :-)

Reply:
`keys` for `Matrix` returns `CartesianIndex` values, which get converted to `Symbol`.

Reply:
OK. Should we really allow this? It doesn't seem to be of any practical use, and it could be confusing (e.g. if you thought you returned a vector but returned a matrix instead).

Reply:
OK. Should we really allow this? It doesn't seem to be of any practical use, and it could be confusing (e.g. if you thought you returned a vector but returned a matrix instead).

Reply:
But then the problem is that we take `keys`. Should we then check if `keys` is either all `Integer` or all `Symbol` and allow only these two cases?

Reply:
Since matrices have a shape, it would make sense to allow returning only single-row matrices with `ByRow`, and generate column names like we do for vectors. What happens when you return a multiple-row data frame with `ByRow`?

Reply:
Since matrices have a shape, it would make sense to allow returning only single-row matrices with `ByRow`, and generate column names like we do for vectors. What happens when you return a multiple-row data frame with `ByRow`?

Reply:
> What happens when you return a multiple-row data frame with ByRow?

1. if it is just `src => ByRow(x -> DataFrame(...))` then nothing - it is just accepted.
2. if it is `src => ByRow(x -> DataFrame(...)) => AsTable` then it is an error, because `DataFrame` does not support `keys`

I think both behaviours are OK.

> Since matrices have a shape, it would make sense to allow returning only single-row matrices with ByRow, and generate column names like we do for vectors.

I would prefer to avoid such a special case for something that is not really needed often. What I expect to be returned most often are: 1) vectors, 2) `Tuple`s, 3) `NamedTuple`s, 4) `Dict`s, 5) `DataFrameRow`. And the rule "use `keys` to get default column names" for them works correctly. I would handle other cases just as an extension of the rule that correctly works for these 5 cases. That is why I proposed to check if `keys` returns `Integer` or `Symbol` (maybe also `AbstractString`) - as these are only useful cases. An alternative to `keys` could be `eachindex` (which for a `Matrix` would give integer indices).

Also if you have a matrix it is easy to wrap it in `vec` to cheaply get a vector without copying anything.

Finally - the fact that matrices have a shape actually should disallow them as in "vector of something" form this "something" should be 1-dimenional (technically we accept an iterable and disregard a dimension).

Reply:
Then what about disallowing returning matrices with `src => ByRow(...) => AsTable` completely? That way we would only accept 1-dimensional or shapeless objects.

Reply:
requiring `keys` to return `Integer` or `Symbol` or `AbstractString` will have exactly this effect (as I think it is an easier rule to understand).

Reply:
I will implement this rule (i.e. restricting allowed types of keys)